### PR TITLE
EE-2404: EE-1120: feat(kingpin) use kingpin to parse ENV vars, and thus allow cli-params _and_ --help

### PR DESCRIPTION
**Pull request by SvenDowideit**
https://portainer.atlassian.net/browse/EE-2404

https://portainer.atlassian.net/browse/EE-1120

(this helps with development of agent features)

we should at _least_ add `--help` - and as kingpin supports Env Vars, we get all the things

```
sven@p1:~/src/portainer/agent$ go run ./cmd/agent/main.go --help
usage: main [<flags>]

Flags:
  --help                        Show context-sensitive help (also try --help-long and --help-man).
  --AgentServerAddr=0.0.0.0     AGENT_HOST address on which the agent API will be exposed
  --AgentServerPort=9001        AGENT_PORT port on which the agent API will be exposed
  --AgentSecurityShutdown=72h   AGENT_SECRET_TIMEOUT the duration after which the agent will be shutdown if not associated or secured by AGENT_SECRET. (defaults to 72h)
  --ClusterAddress=CLUSTERADDRESS  
                                AGENT_CLUSTER_ADDR address (in the IP:PORT format) of an existing agent to join the agent cluster. When deploying the agent as a Docker Swarm service, we can leverage the internal Docker DNS to automatically join existing agents or form a cluster by
                                using tasks.<AGENT_SERVICE_NAME>:<AGENT_PORT> as the address
  --SharedSecret=SHAREDSECRET   AGENT_SECRET shared secret used in the signature verification process
  --LogLevel=INFO               LOG_LEVEL defines the log output verbosity (default to INFO)
  --EdgeMode                    EDGE enable Edge mode. Disabled by default, set to 1 or true to enable it
  --EdgeKey=EDGEKEY             EDGE_KEY specify an Edge key to use at startup
  --EdgeID=EDGEID               EDGE_ID a unique identifier associated to this agent cluster
  --EdgeServerAddr=0.0.0.0      EDGE_SERVER_HOST address on which the Edge UI will be exposed (default to 0.0.0.0)
  --EdgeServerPort=80           EDGE_SERVER_PORT port on which the Edge UI will be exposed (default to 80)
  --EdgeInactivityTimeout="5m"  EDGE_INACTIVITY_TIMEOUT timeout used by the agent to close the reverse tunnel after inactivity (default to 5m)
  --EdgeInsecurePoll            EDGE_INSECURE_POLL enable this option if you need the agent to poll a HTTPS Portainer instance with self-signed certificates. Disabled by default, set to 1 to enable it

```

```
sven@p1:~/src/portainer/agent$ rm /data/agent_edge_key
rm: remove write-protected regular file '/data/agent_edge_key'? y
sven@p1:~/src/portainer/agent$ EDGE=1 EDGE_ID=2fb31223-ddaa-4f95-b2c6-b66ed351050a EDGE_KEY=aHR0cHM6Ly9wb3J0YWluZXIucDEuYWxoby5zdDo5MDAxfHBvcnRhaW5lci5wMS5hbGhvLnN0OjgwMDB8NDM6YjU6YmU6YTQ6N2I6NzI6NGQ6NWE6OTg6MDc6ZDc6ZTM6ODA6ZWI6YmQ6ODJ8MzY  go run ./cmd/agent/main.go 
2022/01/24 17:34:12 [INFO] [main] [message: Agent running on Docker platform]
2022/01/24 17:34:12 [ERROR] [main,docker] [message: Unable to retrieve local agent IP address] [error: Error: No such container: p1]
2022/01/24 17:34:12 [INFO] [main,edge] [message: Edge key loaded from options]
2022/01/24 17:34:12 [INFO] [main,compose] [message: binary is missing, falling-back to compose plugin] [error: docker-compose binary not found]
2022/01/24 17:34:12 [ERROR] [main,edge] [message: Unable to start Edge manager] [error: docker-compose binary not found]
exit status 1
```

seems to be ok (fails getting docker-compose, needs write access to `/data` and other badness when you're not in a container - those assumptions need to both go away, be args, and be documented. (but this is something for another PR :))

```
sven@p1:~/src/portainer/agent$ rm /data/agent_edge_key
rm: remove write-protected regular file '/data/agent_edge_key'? y
sven@p1:~/src/portainer/agent$ go run ./cmd/agent/main.go --EdgeMode --EdgeKey=aHR0cHM6Ly9wb3J0YWluZXIucDEuYWxoby5zdDo5MDAxfHBvcnRhaW5lci5wMS5hbGhvLnN0OjgwMDB8NDM6YjU6YmU6YTQ6N2I6NzI6NGQ6NWE6OTg6MDc6ZDc6ZTM6ODA6ZWI6YmQ6ODJ8MzY --EdgeID=2fb31223-ddaa-4f95-b2c6-b66ed351050a
2022/01/24 17:33:41 [INFO] [main] [message: Agent running on Docker platform]
2022/01/24 17:33:41 [ERROR] [main,docker] [message: Unable to retrieve local agent IP address] [error: Error: No such container: p1]
2022/01/24 17:33:41 [INFO] [main,edge] [message: Edge key loaded from options]
2022/01/24 17:33:41 [INFO] [main,compose] [message: binary is missing, falling-back to compose plugin] [error: docker-compose binary not found]
2022/01/24 17:33:41 [ERROR] [main,edge] [message: Unable to start Edge manager] [error: docker-compose binary not found]
exit status 1
```

### CODE REVIEW

File: cmd/agent/main.go [hunk 1]
```diff
@@ -81,7 +81,8 @@ func main() {
 				log.Printf("[WARN] [main,podman] [message: Unable to retrieve local agent IP address, using '%s' instead] [error: %s]", options.AgentServerAddr, err)
 				advertiseAddr = options.AgentServerAddr
 			} else {
-				log.Fatalf("[ERROR] [main,docker] [message: Unable to retrieve local agent IP address] [error: %s]", err)
+				log.Printf("[ERROR] [main,docker] [message: Unable to retrieve local agent IP address] [error: %s]", err)
+				advertiseAddr = "0.0.0.0"
```
Comment (SvenDowideit) [inline cmd/agent/main.go]: TODO: I was experimenting with non-container based agent - needs more design - will likely move to a different PR

File: os/options.go [hunk 2]
```diff
@@ -34,15 +35,52 @@ func NewEnvOptionParser() *EnvOptionParser {
...
+		EdgeID:                *fEdgeID,
+		EdgeServerAddr:        fEdgeServerAddr.String(), // TODO: really, an agent can't be both edge and non-edge, so we don't need both AgentServerAddr and EdgeServerAddr ?
+		EdgeServerPort:        strconv.Itoa(*fEdgeServerPort),
+		EdgeInactivityTimeout: *fEdgeInactivityTimeout,
+		EdgeInsecurePoll:      *fEdgeInsecurePoll,
+		LogLevel:              *fLogLevel,
+	}, nil
+
+	// obsoleted by kingpin
 	options := &agent.Options{
```
Comment (SvenDowideit) [inline os/options.go]: delete all the code below once i've re-read it

File: cmd/agent/main.go [hunk 3]
```diff
@@ -81,7 +81,8 @@ func main() {
 				log.Printf("[WARN] [main,podman] [message: Unable to retrieve local agent IP address, using '%s' instead] [error: %s]", options.AgentServerAddr, err)
 				advertiseAddr = options.AgentServerAddr
 			} else {
-				log.Fatalf("[ERROR] [main,docker] [message: Unable to retrieve local agent IP address] [error: %s]", err)
+				log.Printf("[ERROR] [main,docker] [message: Unable to retrieve local agent IP address] [error: %s]", err)
+				advertiseAddr = options.AgentServerAddr
```
Comment (SvenDowideit) [inline cmd/agent/main.go]: I think this should be in a different PR - I'm playing with non-container agent use to make my mTLS thing easier to play with

**Comment by SvenDowideit** [comment]
@jamescarppe this may be interesting to you too?

**Comment by SvenDowideit** [comment]
oh, interesting, so there's a bunch of "special" dirs too

```
	HostRoot = "/host"
	// DataDirectory is the folder where the data associated to the agent is persisted.
	DataDirectory = "/data"
	// ScheduleScriptDirectory is the folder where schedules are saved on the host
	ScheduleScriptDirectory = "/opt/portainer/scripts"
	// EdgeKeyFile is the name of the file used to persist the Edge key associated to the agent.
	EdgeKeyFile = "agent_edge_key"
	// DockerBinaryPath is the path of the docker binary
	DockerBinaryPath = "/app"
	// EdgeStackFilesPath is the path where edge stack files are saved
	EdgeStackFilesPath = "/tmp/edge_stacks"
```

ScheduleScriptDirectory seems particularly "interesting"

File: http/handler/browse/browse_delete.go [hunk 4]
```diff
@@ -13,10 +12,6 @@ import (
 // DELETE request on /browse/delete?volumeID=:id&path=:path
 func (handler *Handler) browseDelete(rw http.ResponseWriter, r *http.Request) *httperror.HandlerError {
 	volumeID, _ := request.RetrieveQueryParameter(r, "volumeID", true)
```
Comment (chiptus) [inline http/handler/browse/browse_delete.go:14]: why remove?

Comment (SvenDowideit) [reply http/handler/browse/browse_delete.go:14]: because its not ever set to false

File: internal/edge/key.go [hunk 5]
```diff
@@ -122,3 +134,71 @@ func encodeKey(edgeKey *edgeKey) string {
...
+	var edgeKey string
+
+	edgeKeyFilePath := fmt.Sprintf("%s/%s", agent.DataDirectory, agent.EdgeKeyFile)
+
+	keyFileExists, err := filesystem.FileExists(edgeKeyFilePath)
+	if err != nil {
+		return "", err
+	}
+
+	if keyFileExists {
```
Comment (chiptus) [inline internal/edge/key.go]: shouldn't we send an error if file doesn't exist?

File: internal/edge/key.go [hunk 6]
```diff
@@ -122,3 +134,71 @@ func encodeKey(edgeKey *edgeKey) string {
...
+	}
+
+	return edgeKey, nil
+}
+
+func retrieveEdgeKeyFromCluster(clusterService agent.ClusterService) (string, error) {
+	var edgeKey string
+
+	member := clusterService.GetMemberWithEdgeKeySet()
+	if member != nil {
```
Comment (chiptus) [inline internal/edge/key.go]: shouldn't we fail if member doesn't exist?

File: http/handler/browse/browse_delete.go (hunk 4 above)
Comment (chiptus) [reply http/handler/browse/browse_delete.go:14]: :+1: weird :)

File: internal/edge/key.go (hunk 5 above)
Comment (SvenDowideit) [reply internal/edge/key.go]: idk, i didn't modify the code, i just moved it from main.go to here :)

Comment (SvenDowideit) [reply internal/edge/key.go]: (no, its the "not yet configured case")

File: internal/edge/key.go (hunk 6 above)
Comment (SvenDowideit) [reply internal/edge/key.go]: idk, i didn't modify the code, i just moved it from main.go to here :)

(this one too will then assume that the key comes from the env var)

this code is causing the SetKey weirdness, but i wasn't really wanting to change "everything" :D

**Comment by chiptus** [comment]
> ScheduleScriptDirectory seems particularly "interesting"

is it used anywhere?

File: cmd/agent/main.go (hunk 3 above)
Comment (chiptus) [reply cmd/agent/main.go]: is it in the other PR, if so, should this be reverted?

File: internal/edge/key.go [hunk 7]
```diff
@@ -20,15 +20,26 @@ type edgeKey struct {
 
 // SetKey parses and associates an Edge key to the agent.
 // If the agent is running inside a Swarm cluster, it will also set the "set" flag to specify that a key is set on this agent in the cluster.
+// Don't overwrite the file if it exists - this only appears to work, because our container runs as root
```
Comment (chiptus) [inline internal/edge/key.go]: I think the changes to this file can be saved for the edge refactor PR, no?

and for that, IMO just rewrite the file (and save it as rewritable)

File: cmd/agent/main.go (hunk 3 above)
Comment (deviantony) [reply cmd/agent/main.go]: Probaby should be in another PR

File: os/options.go [hunk 8]
```diff
@@ -16,7 +13,7 @@ const (
 	EnvKeyClusterAddr           = "AGENT_CLUSTER_ADDR"
 	EnvKeyAgentSecret           = "AGENT_SECRET"
 	EnvKeyAgentSecurityShutdown = "AGENT_SECRET_TIMEOUT"
-	EnvKeyCapHostManagement     = "CAP_HOST_MANAGEMENT"
+	//EnvKeyCapHostManagement     = "CAP_HOST_MANAGEMENT"  // deprecated and unused
```
Comment (deviantony) [inline os/options.go]: Maybe remove them instead of commenting them?

File: os/options.go [hunk 9]
```diff
@@ -34,104 +31,39 @@ func NewEnvOptionParser() *EnvOptionParser {
...
+	return &agent.Options{
+		AgentServerAddr:       fAgentServerAddr.String(),
+		AgentServerPort:       strconv.Itoa(*fAgentServerPort),
+		AgentSecurityShutdown: *fAgentSecurityShutdown,
+		ClusterAddress:        *fClusterAddress,
+		SharedSecret:          *fSharedSecret,
+		EdgeMode:              *fEdgeMode,
+		EdgeKey:               *fEdgeKey,
+		EdgeID:                *fEdgeID,
+		EdgeServerAddr:        fEdgeServerAddr.String(), // TODO: really, an agent can't be both edge and non-edge, so we don't need both AgentServerAddr and EdgeServerAddr ?
```
Comment (deviantony) [inline os/options.go]: In Edge mode, there are still 2 HTTP servers running (the agent server "the proxy" and the UI server allowing a user to input an Edge key).

Although they could be served on the same IP/addr for sure, I don't think we had a use case to expose these two over different ip/addr

**Comment by deviantony** [comment]
`ScheduleScriptDirectory` is used by Edge jobs I believe

File: internal/edge/key.go [hunk 10]
```diff
@@ -20,15 +20,26 @@ type edgeKey struct {
...
+	keyRetrieval, _ := retrieveEdgeKeyFromFilesystem()
+	if keyRetrieval == "" {
+		// key not previously saved
+		err = filesystem.WriteFile(agent.DataDirectory, agent.EdgeKeyFile, []byte(key), 0444)
+		if err != nil {
+			return err
+		}
+		keyRetrieval = key
+	}
+	if keyRetrieval != key {
```
Comment (SvenDowideit) [inline internal/edge/key.go]: @deviantony we need to define what should happen if the edge key was stored, but then the user restarts with a different one.

File: cmd/agent/main.go (hunk 3 above)
Comment (deviantony) [reply cmd/agent/main.go]: I've relocated this in https://github.com/portainer/agent/pull/239

File: internal/edge/key.go (hunk 10 above)
Comment (deviantony) [reply internal/edge/key.go]: Yup will give this some consideration and open a separate PR if needed.

**Comment by deviantony** [comment]
@SvenDowideit @chiptus brought a few changes:

* Merged latest version of develop
* Updated base branch to `ce-2.13`
* Relocate non-container code to https://github.com/portainer/agent/pull/239
* Rollback the change to edge key retrieval (to be considered separately)

Give it a review, then let's merge and inform support/platform about these new changes (mostly the ability to use configuration through CLI flags on top of env vars).

**Comment by SvenDowideit** [comment]
yup, very awesome!
